src/astro/databases/aws/redshift.py

```diff
@@ -104,6 +107,96 @@ def load_pandas_dataframe_to_table(
+                conflict_statement += f" AND {stage_table_name}.{conflict_column}={target_table_name}.{conflict_column}"
+        return conflict_statement
+
+    def merge_table(
+        self,
+        source_table: Table,
+        target_table: Table,
+        source_to_target_columns_map: Dict[str, str],
+        target_conflict_columns: List[str],
+        if_conflicts: MergeConflictStrategy = "exception",
+    ) -> None:
+        """
+        Merge the source table rows into a destination table.
+        The argument `if_conflicts` allows the user to define how to handle conflicts.
+
+        :param source_table: Contains the rows to be merged to the target_table
+        :param target_table: Contains the destination table in which the rows will be merged
+        :param source_to_target_columns_map: Dict of target_table columns names to source_table columns names
+        :param target_conflict_columns: List of cols where we expect to have a conflict while combining
+        :param if_conflicts: The strategy to be applied if there are conflicts.
+        """
+        source_schema = source_table.metadata.schema
+        source_table_name = f"{source_schema}.{source_table.name}"
+        target_schema = target_table.metadata.schema
+        target_table_name = f"{target_schema}.{target_table.name}"
+        stage_table_name = random.choice(string.ascii_lowercase) + "".join(
+            random.choice(string.ascii_lowercase + string.digits)
+            for _ in range(UNIQUE_HASH_SIZE - 1)
+        )
+
+        begin_transaction = "BEGIN TRANSACTION"
+        create_temp_table = f"CREATE TEMP TABLE {stage_table_name} (LIKE {target_schema}.{target_table.name})"
+
+        source_columns = list(source_to_target_columns_map.keys())
+        target_columns = list(source_to_target_columns_map.values())
+        target_column_names_string = ",".join(map(str, target_columns))
+        source_column_names_string = ",".join(map(str, source_columns))
+        insert_into_stage_table = (
+            f"INSERT INTO {stage_table_name}({target_column_names_string}) "
+            f"SELECT {source_column_names_string} FROM {source_table_name}"
+        )
+
+        conflict_statement: Optional[str] = self._get_conflict_statement(
+            if_conflicts, stage_table_name, target_table_name, target_conflict_columns
+        )
+
+        insert_into_target_table = (
+            f"INSERT INTO {target_table_name} SELECT * FROM {stage_table_name}"
+        )
+        drop_stage_table = f"DROP TABLE {stage_table_name}"
+        end_transaction = "END TRANSACTION"
+
+        statements = [begin_transaction, create_temp_table, insert_into_stage_table]
+        if conflict_statement:
+            statements.append(conflict_statement)
+        statements.extend([insert_into_target_table, drop_stage_table, end_transaction])
+
+        with self.hook.get_cursor() as cursor:
+            for statement in statements:
+                cursor.execute(statement)
```

Comment:
@pankajkoti Are all the statements executed in the same session? I just want to make sure `BEGIN TRANSACTION` and `END TRANSACTION` are respected when running statements in the different `cursor.execute(statement)` or they should be part of the same query and executed with the single cursor.execute( ";".join(statement)) ??

Reply:
Yes @utkarsharma2 . They are part of the same session and until the point `END TRANSACTION` is executed, nothing is committed to the DB. Also, I had investigated previously on whether we can execute multiple queries in one go, but it is a limitation on Redshift connector.
Ref: https://github.com/aws/amazon-redshift-python-driver/issues/82 & https://github.com/aws/amazon-redshift-python-driver/issues/39